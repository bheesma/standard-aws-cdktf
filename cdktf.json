{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "679d25ee-31c0-4b01-9e78-f9e68dcaa0e4",
  "sendCrashReports": "false",
  "terraformProviders": [
    "hashicorp/aws@~> 4.60.0"
  ],
  "terraformModules": [],
  "context": {
    "excludeStackIdFromLogicalIds": "true",
    "allowSepCharsInLogicalIds": "true"
  }
}