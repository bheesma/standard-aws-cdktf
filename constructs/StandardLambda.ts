import { Construct } from "constructs";
import { LambdaFunction } from "@cdktf/provider-aws/lib/lambda-function";
import { IamRole } from "@cdktf/provider-aws/lib/iam-role";
import { IamRolePolicyAttachment } from "@cdktf/provider-aws/lib/iam-role-policy-attachment";
import * as path from 'path';
import * as fs from 'fs';
import { AssetType, TerraformAsset } from "cdktf";
import { CloudwatchLogGroup } from "@cdktf/provider-aws/lib/cloudwatch-log-group";


export interface StandardLambdaProps {
  lambdaName: string;
}

/// To create AWS Lambda and associate it with other resources in AWS
export class StandardLambda extends Construct {

  readonly function: LambdaFunction;
  readonly role: IamRole;

  constructor(scope: Construct, id: string, props: StandardLambdaProps) {
    super(scope, id);

    // Define policy so IAM role can be assumed by lambda.
    const lambdaRolePolicy = {
      "Version": "2012-10-17",
      "Statement": [
        {
          "Action": "sts:AssumeRole",
          "Principal": {
            "Service": "lambda.amazonaws.com"
          },
          "Effect": "Allow",
          "Sid": ""
        }
      ]
    };

    // New role for Lambda
    this.role = new IamRole(this, props.lambdaName + "-role", {
      name: props.lambdaName + "-role",
      assumeRolePolicy: JSON.stringify(lambdaRolePolicy)
    });

    // Add basic lambda execution policy
    new IamRolePolicyAttachment(this, "lambda-managed-policy", {
      policyArn: 'arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
      role: this.role.name
    });

    // new CloudWatch Log group
    new CloudwatchLogGroup(this, 'lambda-log-group', {
      name: '/aws/lambda/' + props.lambdaName,
      retentionInDays: 30
    });

    // Default code for lambda
    // TODO: Accept the code fiel as parameter.
    const lambdaDefaultcode =
      `
def lambda_handler(event, context):
  print('Hello, world!')
      `

    ///Prepare the lambda code
    const fileName = 'lambda_function.py';
    const folderPath = path.join(process.cwd(), "code");
    const filePath = path.join(folderPath, fileName);

    fs.rmSync(folderPath, { force: true, recursive: true })
    fs.mkdirSync(folderPath, { recursive: true });
    fs.writeFileSync(filePath, lambdaDefaultcode);

    const lambdaCode = new TerraformAsset(this, 'asset', {
      type: AssetType.ARCHIVE,
      path: folderPath
    });

    /// Finally, create the lambda function
    /// TODO: Parameterise as required
    this.function = new LambdaFunction(this, props.lambdaName, {
      functionName: props.lambdaName,
      role: this.role.arn,
      runtime: "python3.9",
      sourceCodeHash: lambdaCode.assetHash,
      filename: lambdaCode.path,
      handler: 'lambda_function.lambda_handler'

    });

  }
}
