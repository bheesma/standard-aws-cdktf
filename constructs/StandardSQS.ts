import { Construct } from "constructs";
import { SqsQueue } from "@cdktf/provider-aws/lib/sqs-queue";
import { LambdaEventSourceMapping } from "@cdktf/provider-aws/lib/lambda-event-source-mapping";
import { StandardLambda } from "./StandardLambda";
import { IamRolePolicyAttachment } from "@cdktf/provider-aws/lib/iam-role-policy-attachment";
import { IamPolicy } from '@cdktf/provider-aws/lib/iam-policy';
import { DataAwsIamPolicyDocument } from '@cdktf/provider-aws/lib/data-aws-iam-policy-document';


export interface StandardSQSProps {
  queueName: string;
}

/// To create SQS queue and associate it with other resources in AWS.
export class StandardSQS extends Construct {

  readonly queue: SqsQueue;
  readonly deadLetterQueue: SqsQueue;

  constructor(scope: Construct, id: string, props: StandardSQSProps) {
    super(scope, id);

    /// Create dead letter queue
    this.deadLetterQueue = new SqsQueue(this, props.queueName + '-dlq', {
      name: props.queueName + '-dlq',
      sqsManagedSseEnabled: true
    })


    /// Create SQS queue
    this.queue = new SqsQueue(this, props.queueName, {
      name: props.queueName,
      sqsManagedSseEnabled: true,
      redrivePolicy: JSON.stringify({
        deadLetterTargetArn: this.deadLetterQueue.arn,
        maxReceiveCount: 5
      })
    });

  }

  /// Function to add SQS event source mapping to Lambda.
  triggerLambda(lambda: StandardLambda, batchSize: number = 10) {

    const lambdaSqsPolicy = new IamPolicy(this, 'sqs-policy', {
      name: `LambdaSQSPolicy`,
      policy: new DataAwsIamPolicyDocument(this, `lambda_sqs_policy`, {
        statement: [
          {
            effect: 'Allow',
            actions: [
              'sqs:SendMessage',
              'sqs:ReceiveMessage',
              'sqs:DeleteMessage',
              'sqs:GetQueueAttributes',
              'sqs:ChangeMessageVisibility',
            ],
            resources: [this.queue.arn],
          },
        ],
      }).json,
    });

    /// Attach policy to lambda role
    const lambdaSqsPolicyAttachment = new IamRolePolicyAttachment(
      this,
      'execution-role-policy-attachment',
      {
        role: lambda.role.name,
        policyArn: lambdaSqsPolicy.arn,
        dependsOn: [lambdaSqsPolicy]
      }
    );

    /// Create event source mapping
    new LambdaEventSourceMapping(this, 'evsourcemap', {
      functionName: lambda.function.functionName,
      eventSourceArn: this.queue.arn,
      batchSize: batchSize,
      dependsOn: [lambdaSqsPolicyAttachment]
    }
    );

  }
}
