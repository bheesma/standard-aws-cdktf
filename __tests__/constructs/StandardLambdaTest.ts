// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import "cdktf/lib/testing/adapters/jest"; // Load types for expect matchers
import { Testing } from "cdktf";
import { StandardLambda } from "../../constructs/StandardLambda";
import { LambdaFunction } from "@cdktf/provider-aws/lib/lambda-function";

describe("StandardLambda Construct", () => {

  const synthedLambda = Testing.synthScope((scope) => {
    new StandardLambda(scope, 'SampleLambda', {
      lambdaName: 'SampleLambda'
    });
  });

  it("generates lambda function with the name", () => {
    expect(synthedLambda).toHaveResourceWithProperties(LambdaFunction, {
      function_name: "SampleLambda"
    });
  })

  it("matches the snapshot", () => {
    expect(synthedLambda).toMatchSnapshot();
  })

});
