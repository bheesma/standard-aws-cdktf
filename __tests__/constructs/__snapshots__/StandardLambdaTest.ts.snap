// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardLambda Construct matches the snapshot 1`] = `
"{
  "resource": {
    "aws_cloudwatch_log_group": {
      "SampleLambda_lambda-log-group_01B09A61": {
        "name": "/aws/lambda/SampleLambda",
        "retention_in_days": 30
      }
    },
    "aws_iam_role": {
      "SampleLambda_SampleLambda-role_5F8F2B8D": {
        "assume_role_policy": "{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Action\\":\\"sts:AssumeRole\\",\\"Principal\\":{\\"Service\\":\\"lambda.amazonaws.com\\"},\\"Effect\\":\\"Allow\\",\\"Sid\\":\\"\\"}]}",
        "name": "SampleLambda-role"
      }
    },
    "aws_iam_role_policy_attachment": {
      "SampleLambda_lambda-managed-policy_76CD1947": {
        "policy_arn": "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
        "role": "\${aws_iam_role.SampleLambda_SampleLambda-role_5F8F2B8D.name}"
      }
    },
    "aws_lambda_function": {
      "SampleLambda_1DD81F2F": {
        "filename": "assets/SampleLambda_asset_2FC1DAF6/A004FBF4B54E411C41147BB16FD0A68B/archive.zip",
        "function_name": "SampleLambda",
        "handler": "lambda_function.lambda_handler",
        "role": "\${aws_iam_role.SampleLambda_SampleLambda-role_5F8F2B8D.arn}",
        "runtime": "python3.9",
        "source_code_hash": "A004FBF4B54E411C41147BB16FD0A68B"
      }
    }
  }
}"
`;
